### DISCUSSION TITLE
Difficulty with parseMultipartData() - SOLVED!

### DISCUSSION THREAD

**Discussion by dirtybirdnj** [strapi/strapi#8428] [Legacy Topics]
Hello, I am trying to write a custom controller endpoint that takes in a CSV file and creates a bunch of records related to a parent record.

**edit: I figured out what was wrong before hitting submit on this post... skip to the end for the answer**

There is another thread related to this, but the details inside it are not fixing the problem so its gotta be something else I am missing

https://github.com/strapi/strapi/discussions/7574

I have isolated / reduced the code in my custom controller down to this snippet, which demonstrates what isn't working properly:

```
async participants(ctx){

    if(ctx.is('multipart')){

        const { data, files } = parseMultipartData(ctx); // files comes back as an empty object
        console.log('multipart', data, files)
        
        if(!files.files) ctx.send('No CSV data / files present.')

    } else {
       console.log('http method error');
       ctx.send(500, 'invalid request type, must be multipart')
    }
    
}

```

I am using the Paw HTTP client to send requests against my API, here is a generated Axios code block that shows the request as it's hitting strapi:

```
axios({
	"method": "PUT",
	"url": "http://localhost:1337/campaigns/3/participants",
	"headers": {
		"Authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MSwiaWF0IjoxNjAzMzE1NTM5LCJleHAiOjE2MDU5MDc1Mzl9.S410vwmmyREJDcQo8AVPug8gzdBtzrT0XBBhIRR9wDY",
		"Content-Type": "multipart/form-data; charset=utf-8; boundary=__X_PAW_BOUNDARY__"
	},
	"data": "--__X_PAW_BOUNDARY__\r\nContent-Disposition: form-data; name=\"files.files\"\r\n\r\nfirstName,lastName,email,phone\r\nArron,Stevens,aaron@email.com,862-266-6666\r\nMike,Smith,mike@email.com,802-867-5309\r\nTeddy,Bushwick,teddy@email.com,802-867-5310\r\nBernie,Sanders,bernie@email.com,802-867-5311\r\nAl,Bundy,al@email.com,802-867-5312\r\n--__X_PAW_BOUNDARY__\r\nContent-Disposition: form-data; name=\"data\"\r\n\r\n{ \"fake\": \"data\"}\r\n--__X_PAW_BOUNDARY__--\r\n"
})
```

Screenshots for anybody familiar:

<img width="1158" alt="Screen Shot 2020-10-21 at 5 39 28 PM" src="https://user-images.githubusercontent.com/419961/96790331-9b013f00-13c4-11eb-8636-0f146911d152.png">

Here's the terminal / server command line output I see when I run the hit the endpoint via Paw, displaying that the `data` part of the multipart form is being parsed but the `files` part comes back as an empty object despite using the undocumented `files.files` labeling attribute.

```
Welcome back!
To manage your project 🚀, go to the administration panel at:
http://localhost:1337/admin

To access the server ⚡️, go to:
http://localhost:1337

multipart { fake: 'data' } {}
[2020-10-21T21:38:53.286Z] debug PUT /campaigns/3/participants (22 ms) 200
```

Things I have tried to make this work:

1. Changing the HTTP verb from POST to PUT (noticed this in the other article)
2. Tried every variation of naming the files attribute (files.files, files.csv, files.files.csv... nothing works)
3. The `data` field IS being picked up though! so what I'm doing isn't 100% off the trail... but something isn't correct.
4. Had another engineer on my team try setting up the same thing via Postman... identical results.
5. Removing either `files` or `data` values from the payload prompts the expected error messages [displayed in the source code](https://github.com/strapi/strapi/blob/master/packages/strapi-utils/lib/parse-multipart.js)

I suspect there is something wrong with the HTTP request I'm sending, but the complete lack of documentation around this functionality is... very frustrating.

If you have encountered this please help! I hope that the knowledge gained in this can be contributed back to the Strapi documentation because it's a vital piece of functionality (mutltipart-form data parsing) that seems to me to be not working at all? I'd be wiling to bet the problem is between the keyboard and the monitor but I'm not gonna figure it out alone. 

Thanks for making Strapi great it's been a phenomenal prototyping tool to work with.

#### EDIT: THE POWER OF RUBBER DUCKING

Nothing better than starting out to write a forum post... only to have the answer fall in your lap. I'm happy to say I know the solution to my problem in case anybody else hits it. I can't say this is a "bug" with strapi (I feel it is but that is not objective)  but I believe at the very least, this needs to be **much** better documented. 

As of today this is the extent of the official documentation I could find covering the subject:

https://strapi.io/documentation/v3.x/concepts/controllers.html#core-controllers

<img width="575" alt="Screen Shot 2020-10-21 at 6 11 51 PM" src="https://user-images.githubusercontent.com/419961/96792742-051be300-13c9-11eb-8dc3-bf6dd33eb584.png">

The problem I had is something to do with the way that Paw / Postman compose their multipart-form payloads. Other times I've seen this `files.files` pattern and it often it requires, for lack of a better way to describe it more unnecessary extraneous fields declared than actual data being passed.

### SOLUTION:
If `files.files` field is not being processed, you need to add an ADDITIONAL `files` field that is also empty

Screenshot:
<img width="1166" alt="Screen Shot 2020-10-21 at 6 06 23 PM" src="https://user-images.githubusercontent.com/419961/96792458-714a1700-13c8-11eb-82d6-2d08a84c1bcd.png">

Here's the Axios request being generated now, compare to above to see the diff:

```
axios({
	"method": "PUT",
	"url": "http://localhost:1337/campaigns/3/participants",
	"headers": {
		"Authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MSwiaWF0IjoxNjAzMzE1NTM5LCJleHAiOjE2MDU5MDc1Mzl9.S410vwmmyREJDcQo8AVPug8gzdBtzrT0XBBhIRR9wDY",
		"Content-Type": "multipart/form-data; charset=utf-8; boundary=__X_PAW_BOUNDARY__"
	},
	"data": "--__X_PAW_BOUNDARY__\r\nContent-Disposition: form-data; name=\"data\"\r\n\r\n{ \"fake\": \"data\"}\r\n--__X_PAW_BOUNDARY__\r\nContent-Disposition: form-data; name=\"files.files\"; filename=\"participants-clean.csv\"\r\nContent-Type: text/csv\r\n\r\nfirstName,lastName,email,phone\r\nArron,Stevens,aaron@email.com,862-266-6666\r\nMike,Smith,mike@email.com,802-867-5309\r\nTeddy,Bushwick,teddy@email.com,802-867-5310\r\nBernie,Sanders,bernie@email.com,802-867-5311\r\nAl,Bundy,al@email.com,802-867-5312\r\n--__X_PAW_BOUNDARY__\r\nContent-Disposition: form-data; name=\"files\"\r\n\r\n\r\n--__X_PAW_BOUNDARY__--\r\n"
})
```

Heres the terminal output now, clearly showing the csv file being passed as desired:

```
[2020-10-21T21:38:53.286Z] debug PUT /campaigns/3/participants (22 ms) 200
multipart { fake: 'data' } {
  files: File {
    _events: [Object: null prototype] {},
    _eventsCount: 0,
    _maxListeners: undefined,
    size: 241,
    path: '/var/folders/xb/jn2mcm7s5sn6hs6n02d16qw40000gn/T/upload_fd7a48435f1ad2f98a86287d63439018',
    name: 'participants-clean.csv',
    type: 'text/csv',
    hash: null,
    lastModifiedDate: 2020-10-21T21:51:18.778Z,
    _writeStream: WriteStream {
      _writableState: [WritableState],
      writable: false,
      _events: [Object: null prototype] {},
      _eventsCount: 0,
      _maxListeners: undefined,
      path: '/var/folders/xb/jn2mcm7s5sn6hs6n02d16qw40000gn/T/upload_fd7a48435f1ad2f98a86287d63439018',
      fd: null,
      flags: 'w',
      mode: 438,
      start: undefined,
      autoClose: true,
      pos: undefined,
      bytesWritten: 241,
      closed: true,
      [Symbol(kFs)]: [Object],
      [Symbol(kCapture)]: false,
      [Symbol(kIsPerformingIO)]: false
    },
    [Symbol(kCapture)]: false
  }
}
```

In conclusion, I don't care how/why it works... just that it finally works! I hope this information can get added to the docs because it's a shame how difficult it was to figure out some pretty simple functionality.

**Comment by derrickmehaffy** [comment]
It's best to move this to our [forums](https://forum.strapi.io/) as we will be getting rid of the discussions (I saw your message on slack)
